Check project item permission and not only global permission

Fearxpc opened the pull request:
The plugin checked only the global permissions.
But often people configure only folder or project permissions to have a more specific access control.
These were ignored.

This works also if a Jenkins instance uses only the global permissions because the project item inherits the permissions.

This commit fixes issue #950

Fearxpc committed:
check project item permission conf instead of the global permssion conf only
replaced deprecated method getInstance with get

bbreijer commented:
Nice one!
I'am not a 100% sure it will lead to the desired result, did you test it on a local instance?

The reason why i'am having doubts is that the given authentication is used to verify the project permission, but this authentication is impersonated using `ACL.SYSTEM` in for instance https://github.com/jenkinsci/gitlab-plugin/blob/e66ccc0fb76d335d5d28945359023db938a6cc24/src/main/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildAction.java#L82-L90

Do you agree?

I'am thinking of using the committers username at that point to check the permissions. @markyjackson-taulia how do you feel about that?

Fearxpc commented:
Thank you for your response.

> I'am not a 100% sure it will lead to the desired result, did you test it on a local instance?

Yes i made a couple of test which were sucessfull.

Created a folder "test" with a project "gitlab_test" inside.
Created dummy user (non admin).

1. Jenkins without any permissions
-> Gitlab triggered the job ✔️ 
2. Jenkins with activated matrix based security (set no permissions)
-> Job was **not** triggered ✔️ 
3. Added Build permission to authenticated users (global permission)
-> Job was triggered ✔️ 
4. Added Build permission to dummy user only (global permission)
-> Job was triggered ✔️ 
5. Added Build permission to dummy user on the folder only
-> Job was triggered ✔️ 
6. Added Build permission to dummy user on the project only
-> Job was triggered ✔️ 


> The reason why i'am having doubts is that the given authentication is used to verify the project permission, but this authentication is impersonated using ACL.SYSTEM in for instance

At the moment I see what you mean but I can't give a good answer to it yet. :) Need to dig a little bit more into the code for it.

Fearxpc commented:
So I made some more tests today and checked the code.

> The reason why i'am having doubts is that the given authentication is used to verify the project permission, but this authentication is impersonated using `ACL.SYSTEM` in for instance
> 
> https://github.com/jenkinsci/gitlab-plugin/blob/e66ccc0fb76d335d5d28945359023db938a6cc24/src/main/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildAction.java#L82-L90

The code impersonates the `ACL.SYSTEM` to call the `TriggerNotifier` method `run()` and the following code but the authentication itself is made with the project ACL object.

https://github.com/jenkinsci/gitlab-plugin/blob/929bb5008866d2cca966cc4e0672420716f350e1/src/main/java/com/dabsquared/gitlabjenkins/webhook/build/BuildWebHookAction.java#L60-L64

`project.getACL()` returns the ACL object of the job/project item and the `authentication` object contains the user which was used to create the webhook in Gitlab.

So yes in all tests I got the expected result :)

bbreijer commented:
Since we are only using the project specific configuration (https://github.com/jenkinsci/gitlab-plugin/blob/e66ccc0fb76d335d5d28945359023db938a6cc24/README.md#L121), i did not consider the configuration mentioned in  (https://github.com/jenkinsci/gitlab-plugin/blob/e66ccc0fb76d335d5d28945359023db938a6cc24/README.md#L114). For this configuration, your PR works perfectly!

I tested everything using the `Global configuration` method, and it makes perfect sense and there are no doubts left for me ;)
